File: REST/src/main/java/com/tris/background/Announcer.java
package com.tris.background;

import java.util.Calendar;
import java.util.Random;
import java.util.TimeZone;
import java.util.concurrent.TimeUnit;

public class Announcer extends Thread {

	private AliveAnnouncer alive;
	private ByeByeAnnouncer bye;
	private Thread timer;
	private Runnable sendRun;

	public Announcer() {
		Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC+1"));
		long bootid = calendar.getTimeInMillis() / 1000L;

		alive = new AliveAnnouncer(bootid);
		bye = new ByeByeAnnouncer(bootid);
		
		sendRun = new Runnable() {
			
			@Override
			public void run() {
				onEvent("alive");
			}
		};
		
		timer = new Thread(new Runnable() {

			@Override
			public void run() {
				Random r = new Random();
				while (true) {
					try {
						int timeout = r.nextInt(AliveAnnouncer.EXPIRATION_TIME / 2);
						System.out.println("Next alive msg in: " + timeout + " seconds");
						TimeUnit.SECONDS.sleep(timeout);
					} catch (InterruptedException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}

					new Thread(sendRun).start();
				}
			}
		});

	}
	
	@Override
	public void run() {
		new Thread(sendRun).start();
		timer.start();
	}

	private void onEvent(String type) {
		switch (type) {
		case "alive":
			try {
				alive.sendAliveMessageBundle();
			} catch (Exception e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			break;

		case "shutdown":
			timer.stop();
			try {
				bye.sendByeByeMessageBundle();
			} catch (Exception e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			break;

		default:
			break;
		}
	}

	public void shutdown() {
		onEvent("shutdown");
	}

}

File: Real Device/api/background/Announcer.py
from threading import Thread, Lock
import random
import time

from AliveAnnouncer import AliveAnnouncer
from ByeByeAnnouncer import ByeByeAnnouncer

import logging

class Announcer(Thread):

    ALIVE_TYPE = 'alive'
    SHUTDOWN_TYPE = 'shutdown'

    shutdownLock = Lock()
    shutdownFlag = False

    def __init__(self, expirationTime, bootid):
        Thread.__init__(self)

        self.expirationTime = expirationTime
        self.bootid = bootid

        self.alive = AliveAnnouncer(expirationTime, bootid)
        self.bye = ByeByeAnnouncer(bootid)

    def run(self):
        logging.info('Announcer started')
        Thread(target=self.onEvent, args=(self.ALIVE_TYPE,)).start()
        Thread(target=self.timer).start()

    def timer(self):
        while True:
            timeout = random.randint(0, self.expirationTime / 2)
            logging.info('Next alive msg in: %d' % (timeout))
            time.sleep(timeout)

            self.shutdownLock.acquire()
            if not self.shutdownFlag:
                Thread(target=self.onEvent, args=(self.ALIVE_TYPE,)).start()
                self.shutdownLock.release()
            else:
                self.shutdownLock.release()
                logging.debug('Finishing Timer due to shutdown')
                break

    def onEvent(self, type):
        if type == self.ALIVE_TYPE:
            self.alive.sendAliveMessageBundle()
        else:
            self.shutdownLock.acquire()
            self.shutdownFlag = True
            self.bye.sendByeByeMessageBundle()
            self.shutdownLock.release()

    def shutdown(self):
        self.onEvent(self.SHUTDOWN_TYPE)

if __name__ == '__main__':
    logging.basicConfig(level=logging.DEBUG)
    a = Announcer(10, 1)
    a.start()
    time.sleep(15)
    a.shutdown()
